Standardize node ops sync_nodes selection

bhalevy opened the pull request:
Use token_metadata get_endpoint_to_host_id_map_for_reading
to get all normal token owners for all node operations,
rather than using gossip for some operation and
token_metadata for others.

Fixes #12862

bhalevy commented:
```
[gw2] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_removenode_rejected_with_rnbo 
[gw3] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_lcs_reshape_efficiency 
[gw5] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_enable_rbno_for_bootstrap 
[gw1] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_enable_rbno_for_default_operation 
[gw0] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_disable_rbno 
[gw4] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_disable_rbno_for_all_operations 
[gw0] PASSED repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_ignore_dead_nodes_for_replace_option 
```

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/167/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/178/

asias commented on service/storage_service.cc:
```diff
@@ -629,18 +629,18 @@ future<> storage_service::mark_existing_views_as_built(sharded<db::system_distri
+                ignore_nodes.insert(ep_and_id.endpoint);
             }
         } catch (...) {
             throw std::runtime_error(format("Failed to parse --ignore-dead-nodes-for-replace parameter: ignore_nodes={}, node={}: {}", ignore_nodes_strs, n, std::current_exception()));
```
The comment for "replace" is not updated after the renaming of the function.

asias commented on service/storage_service.cc:
```diff
@@ -2248,21 +2261,15 @@ void storage_service::run_bootstrap_ops(std::unordered_set<token>& bootstrap_tok
+
+    // Step 1: Decide who needs to sync data for bootstrap operation
     // TODO: Specify ignore_nodes
+    ctl.start("bootstrap", [this_node = get_broadcast_address()] (gms::inet_address node) {
```
We do not need to exclude this_node. Original code excludes this_node because we do not bother to check the gossip status of this_node.  In ctrl.start, get_endpoint_to_host_id_map_for_reading will not return this_node any way. 

Note, at line 2274, we insert this_node to sync_nodes explicitly. 

It is better to drop the this_node check filter here to avoid confusion.

bhalevy replied:
okay, makes sense

bhalevy replied:
Actually, it turns out that get_endpoint_to_host_id_map_for_reading does return pending nodes.
It's not restricted to normal town owners.
I'll delete the first patch then.
I'm tempted to change refresh_sync_nodes to return only normal token owners
but we should do that as a followup since it would impact decommission and removenode.

asias commented on service/storage_service.cc:
```diff
@@ -2288,11 +2295,13 @@ void storage_service::run_bootstrap_ops(std::unordered_set<token>& bootstrap_tok
             slogger.warn("bootstrap[{}]: Found pending node ops = {}, sleep 5 seconds and check again", uuid, pending_ops);
             sleep_abortable(std::chrono::seconds(5), _abort_source).get();
+
+            ctl.refresh_sync_nodes([this_node = get_broadcast_address()] (gms::inet_address node) {
```
We can drop the filter here too.

But refresh_sync_nodes will revert what we do at line 2274 which adds this_node to sync_nodes. 

We need to call here:

```
ctl.refresh_sync_nodes();
ctl.sync_nodes.insert(get_broadcast_address());
```

bhalevy replied:
But it will be re-added by line 2274 when we loop back.

bhalevy commented:
In v2 ([dc6e9e6](https://github.com/scylladb/scylladb/commit/dc6e9e6776850853061acc0d71d38a2889e4c499)):
- Reverted "token_metadata: get_endpoint_to_host_id_map_for_reading: document return of normal token owners"
- storage_service: get_ignore_dead_nodes_for_replace: make static and rename to parse_node_list
  - renamed function to a more generic name
  - and updated log printout
- storage_service: node ops: standardize sync_nodes selection
  - erase the bootstrapping node from sync_nodes and add it back explicitly after `wait_for_normal_state_handled_on_boot`  for bootstrap and replace.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/216/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/337/

bhalevy commented:
v3 ([ca61d88](https://github.com/scylladb/scylladb/commit/ca61d88764f234851c95a26123dbe72abd28ed4b)):
- rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/427/

bhalevy commented:
@asias please ack.

bhalevy commented:
ping @asias

asias commented:
> ping @asias

Sorry, I was out last week. 

Looks good to me.

bhalevy commented:
@tgrabiec / @scylladb/scylla-maint please merge
